pkg/cloudprovider/provider/vmware-cloud-director/provider.go

```diff
@@ -0,0 +1,531 @@
+		return nil, err
+	}
+
+	return p.getInstance(vm)
+}
+
+func (p *provider) GetCloudConfig(spec clusterv1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) getConfig(provSpec clusterv1alpha1.ProviderSpec) (*Config, *providerconfigtypes.Config, *vcdtypes.RawConfig, error) {
+	if provSpec.Value == nil {
+		return nil, nil, nil, fmt.Errorf("machine.spec.providerconfig.value is nil")
+	}
+
+	pconfig, err := providerconfigtypes.GetConfig(provSpec)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if pconfig.OperatingSystemSpec.Raw == nil {
+		return nil, nil, nil, errors.New("operatingSystemSpec in the MachineDeployment cannot be empty")
+	}
+
+	rawConfig, err := vcdtypes.GetConfig(*pconfig)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	c := Config{}
+	c.Username, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.Username, "VCD_USER")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"username\" field, error = %w", err)
+	}
+
+	c.Password, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.Password, "VCD_PASSWORD")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"password\" field, error = %w", err)
+	}
+
+	c.Organization, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.Organization, "VCD_ORG")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"organization\" field, error = %w", err)
+	}
+
+	c.URL, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.URL, "VCD_URL")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"url\" field, error = %w", err)
+	}
+
+	c.VDC, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.VDC, "VCD_VDC")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"vdc\" field, error = %w", err)
+	}
+
+	c.AllowInsecure, err = p.configVarResolver.GetConfigVarBoolValueOrEnv(rawConfig.AllowInsecure, "VCD_ALLOW_UNVERIFIED_SSL")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"allowInsecure\" field, error = %w", err)
+	}
+
+	c.VApp, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.VApp)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	c.Template, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Template)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	c.Catalog, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Catalog)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	c.Network, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Network)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	c.IPAllocationMode = rawConfig.IPAllocationMode
+	c.CPUs = rawConfig.CPUs
+	c.CPUCores = rawConfig.CPUCores
+	c.MemoryMB = rawConfig.MemoryMB
+	c.DiskSizeGB = rawConfig.DiskSizeGB
+	c.DiskBusType = rawConfig.DiskBusType
+	c.DiskIOPS = rawConfig.DiskIOPS
```

Comment:
Are there any minimum and maximum values for those fields? For example, what if someone provides 0 IOPS? Or even something like negative IOPS or size?

Reply:
0 IOPS is allowed and the default value. Yeah maybe we can check for negative values :)